### PR TITLE
Limited Kusama <> Polkadot bridge

### PR DESCRIPTION
**Pull request by svyatonik**
Even though there are some TODOs here, the PR is reviewable already (the only **planned** code changes are weight-related). Main TODOs in this PR:
- [x] deal with weights - that's the first time we'll be adding bridge benchmarks to the polkadot runtimes, so something may go wrong;
- [ ] upgrade - do we need it at all, or we should use governance process for that?
- [ ] get an explicit approval from @gavofyork on main use-case. I'm not sure if this has been discussed in details before - all that is presented here is just my vision and I have a limited knowledge of Kusama/Polkadot governance mechanisms;
- [ ] security audit of the integration code. All bridge pallets are already audited, but there's some recent updates in the Kusama and Polkadot runtimes that need an audit (all commits, starting from 7f0a28d447de0c270e754397279686a63f0fa4a5 need an audit).

## Short bridge FAQ

This bridge adds Polkadot GRANDPA light client (`pallet-bridge-gandpa`) to the Kusama runtime and vice versa. This light client may be used to verify proofs (e.g. storage proofs) from the bridged chain. In our implementation, we'll be accepting *messages* (using `pallet-bridge-messages`) at the *source chain* and submit storage proofs of these messages to the *target chain*. Both chains may serve as the *source* and *target* chains. The *message* in our implementation means encoded call (that's the main part of the regular transaction) that will be accepted at the *source chain* and then dispatched (using `pallet-bridge-dispatch`) at the *target chain*.

Apart from the on-chain pallets, there's also offchain entity - *relay*/*relayer*, which is submitting message delivery transactions to both chains.

More details can be found in the [`parity-bridges-common`](https://github.com/paritytech/parity-bridges-common) repository.

## Bridge limitations

While the bridge we develop is a general-usage bridge, it'll have serious restrictions in initial deployment.

1. Bridge may only be used by the Kusama council
2. Only `balances::transfer` calls may be sent over the bridge
3. The bridge will be slow - e.g. it may take 1 hour to deliver message from Kusama to Polkadot and 4 hours to deliver message from Polkadot to Kusama

Why is it so restricted? The main reason is that the bridge is going to change in the future (detailed plan is still WiP) and if we'll allow everyone to use it now, it'll be hard, if not impossible, to perform migrations.

## Bridge use-case

@gavofyork - This is the most questionable part here, where I'd like to get your approval.

How it is designed:
1) there'll be some Kusama account that will pay for the Kusama council messages - `KusamaCouncilAccount`;
2) once bridge is deployed, we'll configure it to use the this account for payments (by setting `AllowedMessageSender` parameter from the `polkadot_messages.rs` file);
3) we have a means to 'derive' accounts at the target chain from accounts of the source chain. In our case, we'll need to derive Polkadot account from the `KusamaCouncilAccount`. Let it be the `KusamaCouncilAccountAtPolkadot`;
4) at Polkadot, DOTs may be sent to this accounts from the treasury;
5) Kusama council members may submit send-message-over-bridge motions, thus spending (transferring) DOTs from `KusamaCouncilAccountAtPolkadot` at the Polkadot.

Technically it means, that we'll only accept messages that are being sent using origin `pallet_collective::RawOrigin::<CouncilCollective>::Members(_, _)` or `pallet_collective::RawOrigin::<CouncilCollective>::Member()`. So even one member can send messages (aka spend DOTs)!

Please check if I'm wrong here - I guess we may need some corrections here.

## Local Demo

Everyone may test it with local Kusama/Polkadot chains, using [this guide](https://hackmd.io/@FECu6P3WSzmvGvwaLcf7KA/r1nCHGUG9). This guide needs some updates - in particular, how to fill all send-message transaction parameters, but overall it is usable.

There's also [kusama-polkadot.test](https://github.com/svyatonik/kusama-polkadot.test) repository, which may be used for spinning up local Kusama and Polkadot nodes + running the bridge. But it is a bit outdated and maintenance is not guaranteed.

P.S.: I'm pinging @andresilva as well, as suggested by @stefan-sopic, since you're somewhat related to bridges :) Feel free to ignore if you feel its outside of your interests/expertise.

### CODE REVIEW

File: runtime/kusama/src/lib.rs [hunk 1]
```diff
@@ -1541,6 +1637,56 @@ impl OnRuntimeUpgrade for CrowdloanIndexMigration {
 	}
 }
 
+/// Initialize with-Polkadot bridge.
+pub struct InitializePolkadotBridge;
```
Comment (svyatonik) [inline runtime/kusama/src/lib.rs:1641]: I'm unsure about this upgrade. We'll need to initialize the "pallet owner" field of both bridge pallets (that's the account that may issue some special "requests" - e.g. halt bridge, update conversion rate, ...). Don't know if doing this during runtime upgrade or using some governance mechanisms after runtime upgrade is better here.

File: runtime/kusama/src/polkadot_messages.rs [hunk 2]
```diff
@@ -0,0 +1,620 @@
...
+
+impl SenderOrigin<AccountId> for Origin {
+	fn linked_account(&self) -> Option<AccountId> {
+		match self.caller {
+			// in benchmarks we accept messages from regular users
+			#[cfg(feature = "runtime-benchmarks")]
+			crate::OriginCaller::system(frame_system::RawOrigin::Signed(ref submitter)) =>
+				Some(submitter.clone()),
+
+			OriginCaller::Council(_) => AllowedMessageSender::get(),
```
Comment (svyatonik) [inline runtime/kusama/src/polkadot_messages.rs:350]: That's another questionable part, which I've mentioned in description. It means that we'll only accept messages with `OriginCaller::Council` origin

**Comment by gavofyork** [comment]
So this isn’t quite about allowing the Kusama *Council* to spend Kusama’s DOT. And it’s certainly not about allowing any individual council member to spend the DOT. Rather it’s about allowing the Kusama *treasury* to own DOT. In the light of Gov2, which will be rolled out imminently this distinction is especially important since there will not be a council if/when it gets deployed.

**Comment by svyatonik** [comment]
> So this isn’t quite about allowing the Kusama Council to spend Kusama’s DOT. Rather it’s about allowing the Kusama treasury to own DOT.

Ok, I may be wrong here as I've already lost origins of this requirement. But it is what we/I've been working here so far. Possibly, it was just one possible implementations:
1) Kusama council is responsible for approving/rejecting treasury funds spending requests. I.e. it kinda "controls" treasury;
2) we are "adding DOTs to Kusama treasury" by allowing council to control DOTs on Polkadot through the bridge.

The original idea was to have some kind of wrapped DOTs on Kusama (iirc some pallet that implements `Currency` trait or just some of its APIs and sends bridge messages when is called) was dropped in favor of this ^^^ at some point.

> And it’s certainly not about allowing any individual council member to spend the DOT.

Yes - that's what I'm unsure about. The options are to change it to majority or complete council. But it depends on whether all this schema works (where council directly controls DOTs over the bridge) for you.

> In the light of Gov2, which will be rolled out imminently this distinction is especially important since there will not be a council if/when it gets deployed.

If there won't be a council or any other group (read: dispatch origin) that we may grant control over bridged DOTs, then we'll need to be able to process spend-treasury-DOTs somehow, right? If we'll keep this control-DOTs-over-bridge concept, then there's a problem even at first glance. Right now we're storing `weight` and `spec_version` with the encoded call in the bridge message and while referendum is active (which I suppose takes some time), the message could "spoil" by the time it'll be dispatched. This could happen if bridged runtime (Polkadot in our case) is upgraded because of weight and version mismatch. I guess XCM could help here, but its integration isn't even started. We can add other intermediate stable format for that that will only support `Balances::transfer` calls, though.

---

In any case, the future of this PR isn't clear to me. Shall we introduce some minor fixes and make it work with majority/complete council approval (mean: governance v1)? Shall we postpone bridge and start adapting it for governance v2? Do we need to change the way how Kusama council/members can control DOTs - i.e. wDOTs instead of control-DOTs-over-bridge?

**Comment by svyatonik** [comment]
the decision is to deploy XCM bridge (either relay or para) instead. closing

**Pull request closed without merging**